pkg/chain/bitcoin/derive_address.go

```diff
@@ -123,27 +122,21 @@ func deriveAddress(extendedPublicKey string, addressIndex uint32) (string, error
 	return finalAddress.EncodeAddress(), nil
 }
 
-// ResolveAddress resolves a configured beneficiaryAddress into a
-// valid bitcoin address. If the supplied address is already a valid bitcoin
-// address, we don't have to do anything. If the supplied address is an
-// extended public key of a HD wallet, attempt to derive the bitcoin address at
-// the specified index.
-func ResolveAddress(
-	beneficiaryAddress string,
-	storage *DerivationIndexStorage,
-	chainParams *chaincfg.Params,
-	handle bitcoin.Handle,
-) (string, error) {
-	// If the address decodes without error, then we have a valid bitcoin
-	// address. Otherwise, we assume that it's an extended key and we attempt to
-	// derive the address.
-	_, err := btcutil.DecodeAddress(beneficiaryAddress, chainParams)
+// ValidateAddress checks to see if the supplied btc address is valid on the
+// supplied chain. We check both raw btc addresses and *pub extended keys.
+func ValidateAddress(btcAddress string, chainParams *chaincfg.Params) error {
+	_, err := btcutil.DecodeAddress(btcAddress, chainParams)
```

Comment:
We coul validate the address for chain like this:
```go
	decodedAddress, err := btcutil.DecodeAddress(btcAddress, chainParams)
	if err != nil {
		...
	}

	if !decodedAddress.IsForNet(chainParams) {
		return fmt.Errorf(
			"provided address [%s] is not a valid btc address for chain [%s]",
			btcAddress,
			chain.Name(),
		)
	}
```

Reply:
Is there a case where the address will properly decode, but `IsForNet` will be false? Adding that in didn't change any of my test results (maybe there's a gap in the tests)

Reply:
Yes, it should fail if you provide e.g. mainnet address and testnet chain params

Reply:
I added some mismatched address/chain tests in da3b5ef9ce490e9643115d0a78a9698dde8dc8a9 that appear to be working, though the error message isn't the greatest

Reply:
btcutil.DecodeAddress is not using provided chain params to validate the address but it's using the chain params to generate an address in case it receives a public key.

Reply:
Sorry - to elaborate a little: 

`1MjCqoLqMZ6Ru64TTtP16XnpSdiE8Kpgcx` is a mainnet address.

If I run 
```go
btcutil.DecodeAddress("1MjCqoLqMZ6Ru64TTtP16XnpSdiE8Kpgcx", &chaincfg.TestNet3Params)
```
I get `error parsing extended public key: [the provided serialized extended key length is invalid]`

How do you want to handle that?

Reply:
What  do you think about including both errors in the final validation error message:
```go
func ValidateAddress(btcAddress string, chainParams *chaincfg.Params) error {
	decodedAddress, decodeErr := btcutil.DecodeAddress(btcAddress, chainParams)
	if decodeErr != nil {
		_, deriveErr := DeriveAddress(btcAddress, 0)
		if deriveErr != nil {
			return fmt.Errorf(
				"[%s] is not a valid btc address using chain [%s]: "+
					"decode address failed with [%v] and "+
					"derive address failed with [%v]",
				btcAddress,
				chainParams.Name,
				decodeErr,
				deriveErr,
			)
		}
		return nil
	}

	if !decodedAddress.IsForNet(chainParams) {
		return fmt.Errorf(
			"provided address [%s] is not a valid btc address for chain [%s]",
			btcAddress,
			chainParams.Name,
		)
	}

	return nil
}
```